tests/rbac_test.go

```diff
@@ -0,0 +1,213 @@
+		By(fmt.Sprintf("Add %s to user %s", role, ktests.UsernameTestUser))
+		args := []string{"adm", "policy", "add-role-to-user", role, ktests.UsernameTestUser, "-n", namespace}
+		_, err := tests.RunCommand("oc", args...)
+		Expect(err).ToNot(HaveOccurred())
+
+		By(fmt.Sprintf("Login with the user %s", ktests.UsernameTestUser))
+		args = []string{"login", "-u", ktests.UsernameTestUser, "-p", "123456"}
+		_, err = tests.RunCommand("oc", args...)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Create a VM with yaml.")
+		args = []string{"create", "-f", vmi.Manifest, "-n", namespace}
+		_, err = tests.RunCommand("oc", args...)
+		if con == true {
+			Expect(err).ToNot(HaveOccurred())
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+
+		By("Check VMI should create.")
+		args = []string{"start", vmi.Name, "-n", namespace}
+		_, err = tests.RunCommand("virtctl", args...)
+		if con == true {
+			Expect(err).ToNot(HaveOccurred())
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+
+		By("Check the vmi is running.")
+		args = []string{"get", "vmi", vmi.Name, "--template", "{{.status.phase}}", "-n", namespace}
+		result := Eventually(func() string {
+			output, err := tests.RunCommand("oc", args...)
+			if con == true {
+				Expect(err).ToNot(HaveOccurred())
+			} else {
+				Expect(err).To(HaveOccurred())
+			}
+			return output
+		}, time.Minute*10)
+		if con == true {
+			result.Should(Equal("Running"))
+		}
+
+		By("Connect via console.")
+		vm, err := virtClient.VirtualMachineInstance(namespace).Get(vmi.Name, &metav1.GetOptions{})
+		if con == true {
+			expecter, err := tests.LoggedInCirrosExpecter(vm)
+			Expect(err).ToNot(HaveOccurred())
+			defer expecter.Close()
+		} else {
+			_, err = tests.LoggedInCirrosExpecter(vm)
+			Expect(err).To(HaveOccurred())
+		}
+
+		By("Checking the response from VNC server.")
+		response, err := vncConnection(namespace, vmi.Name, con)
+		if con == true {
+			Expect(response).To(Equal("RFB 003.008"))
+			Expect(err).ToNot(HaveOccurred())
+		} else {
+			Expect(response).ToNot(Equal("RFB 003.008"))
+			Expect(err).To(HaveOccurred())
+		}
+
+		By("Check VMI should stop.")
+		args = []string{"stop", vmi.Name, "-n", namespace}
+		_, err = tests.RunCommand("virtctl", args...)
+		if con == true {
+			Expect(err).ToNot(HaveOccurred())
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+
+		By("Delete a VM with yaml.")
+		args = []string{"delete", "-f", vmi.Manifest, "-n", namespace}
+		_, err = tests.RunCommand("oc", args...)
+		if con == true {
+			Expect(err).ToNot(HaveOccurred())
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+	},
+		Entry("with admin permission", "admin", tests.NamespaceTestDefault, true),
+		Entry("with edit permission", "edit", tests.NamespaceTestDefault, true),
+		Entry("with view permission", "view", tests.NamespaceTestAlternative, false))
+})
+
+func vncConnection(namespace, vmname string, con bool) (string, error) {
```

Comment:
I noticed there's a similar function in pr https://github.com/kubevirt/kubevirt-ansible/pull/395/files#diff-8337e1399854c6a050cd98f558f9ce68R168, is there any way we could write one generic vnc/serial console function for both scenarios. 
avoid adding `con bool` in parameter, could we check errors outside the function instead ?

Reply:
done

Reply:
@shiywang, I wrote the generic function for my requirement and removed the `con` parameter.